src/queries.ts

```diff
@@ -0,0 +1,102 @@
+import * as vscode from "vscode";
+import {
+  type Model,
+  quickInfoRequest,
+  createInlayHint,
+  getLeftMostHintOfLine
+} from "./helpers";
+
+/** Strongly-typed RegExp groups (https://github.com/microsoft/TypeScript/issues/32098#issuecomment-1279645368) */
+type RegExpGroups<T extends string> = (
+  & IterableIterator<RegExpMatchArray>
+  & {
+    groups: { [name in T]: string } | { [key: string]: string };
+  }[]
+);
+
+const twoslashQueryRegex = /^\s*\/\/\s*\^\?/gm; // symbol: ^?
+const inlineQueryRegex = /^[^\S\r\n]*(?<start>\S).*\/\/\s*(?<end>=>)/gm; // symbol: =>
+type InlineQueryMatches = RegExpGroups<"start" | "end">;
+
+type Query = {
+  text: string;
+  offset: number;
+  model: Model;
+  cancel: vscode.CancellationToken;
+};
+
+type InlayHintsPromise = Promise<vscode.InlayHint[]>;
+
+/** Checks for and returns regular two-slash query hints (symbol: ^?). */
+async function checkTwoslashQuery({ text, offset, model, cancel }: Query): InlayHintsPromise {
+  const results: vscode.InlayHint[] = [];
+  const m = text.matchAll(twoslashQueryRegex);
+
+  for (const match of m) {
+    if (match.index === undefined) {
+      break;
+    }
+
+    const end = match.index + match[0].length - 1;
+    // Add the start range for the inlay hint
+    const endPos = model.positionAt(end + offset);
+
+    if (cancel.isCancellationRequested) {
```

Comment:
Can the cancellation token change in the middle of the for loop? If not, this check could be moved to `getHintsFromQueries()`